Util/channel/src/main/java/io/deephaven/util/channel/BaseSeekableChannelContext.java

```diff
@@ -8,29 +8,36 @@
 import org.jetbrains.annotations.Nullable;
 
 import javax.annotation.OverridingMethodsMustInvokeSuper;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.function.Supplier;
 
 public class BaseSeekableChannelContext implements SeekableChannelContext {
     /**
-     * An opaque resource object hosted by this context.
+     * A cache of opaque resource objects hosted by this context.
      */
-    private SafeCloseable resource;
+    private final Map<String, SafeCloseable> resourceCache = new HashMap<>();
```

Comment:
This would be good candidate to apply a typed key.

```java
interface Key<T> {}

public final <T extends SafeCloseable> T apply(Key<T> key, Supplier<T> factory) { ... }
```

```java
private static final Key<BytesArrayHolder> BUFFER_HOLDER_KEY = new Key<> { };
...
```

a similar pattern is used with `io.grpc.Metadata.Key`.

Reply:
I very much want to do that. The challenge is that I don't have a good place to put the `Key` interface.
The module extensions-parquet-compression (which has all the decompressor stuff) depends on very few modules, and I want to avoid adding extra dependencies.

Reply:
This suggests to me that `BiFunction<String, Supplier<SafeCloseable>, SafeCloseable>` / change to `io.deephaven.parquet.compress.CompressorAdapter#decompress` is too generic and our dependencies between these different modules are not well defined.

Reply:
I've pushed a branch devin-sm-stream-decom illustrating how we might want to re-org this.